ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -144,225 +236,283 @@ chainSyncClient
+      -- server/candidate is on a fork that forked off too far in the past so
+      -- that we do not intersect with them within the last @k@ blocks, so we
+      -- don't want to sync with them.
+      unless (AF.withinFragmentBounds genesisPoint currentFrag) $
+        disconnect ForkTooDeep
+          { _intersection = genesisPoint
+          , _ourTip       = ourTip
+          , _theirTip     = theirTip
+          }
 
       -- Get the 'ChainState' at genesis.
-      let candidateChain' = Empty genesisPoint
-      candidateChainState' <- case rewindChainState cfg curChainState Origin of
-        Nothing -> disconnect $ ForkTooDeep genesisPoint theirHead
-        Just c  -> pure c
-
-      writeTVar varCandidate CandidateState
-        { candidateChain       = candidateChain'
-        , candidateChainState  = candidateChainState'
-        }
-
-      ourHeadBlockNo <- getTipBlockNo
-
-      return $ requestNext mkPipelineDecision0 Zero ourHeadBlockNo theirHead
-
-    requestNext :: MkPipelineDecision
+      candidateChainState <-
+        case rewindChainState cfg currentChainState Origin of
+          Nothing -> disconnect ForkTooDeep
+            { _intersection = genesisPoint
+            , _ourTip       = ourTip
+            , _theirTip     = theirTip
+            }
+          Just c  -> return c
+
+      let is = IntersectionState
+            { candidateFrag       = Empty genesisPoint
+            , candidateChainState = candidateChainState
+            , currentFrag         = currentFrag
+            , currentTip          = ourTip
+            }
+
+      writeTVar varCandidate $ Empty genesisPoint
+
+      nextStep is mkPipelineDecision0 Zero theirTip
+
+    -- | Look at the current chain fragment that may have been updated in the
+    -- background. Check whether the candidate fragment still intersects with
+    -- it. If so, update the 'IntersectionState' and trim the candidate
+    -- fragment to the new current chain fragment's anchor point. If not,
+    -- return 'Nothing'.
+    intersectsWithCurrentChain
+      :: IntersectionState blk tip
+      -> STM m (Maybe (IntersectionState blk tip))
+    intersectsWithCurrentChain is@IntersectionState { candidateFrag, currentFrag } = do
+      currentFrag' <- getCurrentChain
+      currentTip'  <- Our <$> getOurTip
+      if
+        | AF.headPoint currentFrag == AF.headPoint currentFrag ->
+          -- Our current chain didn't change, nothing to do
+          return $ Just is
+
+        | isJust (AF.intersectionPoint currentFrag' candidateFrag) ->
+          -- Our current chain changed, but it still intersects with candidate
+          -- fragment, so update the 'currentFrag' field and trim to the
+          -- candidate fragment to the same anchor point.
+          case AF.splitAfterPoint candidateFrag (AF.anchorPoint currentFrag') of
+           Nothing -> error
+             "anchor point must be on candidate fragment if they intersect"
```

Comment:
Let's motivate this:

* Before the update to our fragment, both fragments were anchored at the same anchor.
* We still have an intersection. 
* The number of blocks after the intersection cannot have shrunk, but could have increased.
* If it did increase, the anchor point will have shifted up.
* It can't have moved up past the intersection point (because then there would be no intersection anymore).
* This means the new anchor point must be between the old anchor point and the new intersection point.
* Since we know both the old anchor point and the new intersection point exist on their fragment, the new anchor point must also.